### DISCUSSION TITLE
A way to force controller usage, i.e. ignoring hand tracking?

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#363] [Q&A]
I'm getting less and less enthousiastic with the hand tracking on our XR-3, especially for demos by VR novices it really doesn't work very well. I searched for a flag in SK to ignore native hand tracking and use detected controllers instead, but I suspect such a flag isn't currently available?

**Comment by maluoi** [comment]
There is not currently, but that sounds like a good feature to have!

Out of curiosity, how does the XR-3 system generally behave when controllers are present? Does it disable hand-tracking and switch to controllers automatically or is that a manual process? Or is it neither of those, and just track hands and controllers both at the same time?

Also curious, how does XR-3's hand tracking stack up against Quest, if you know? Just wondering if this would be a common request regardless of headset for those demoing to new VR people.

(I also opened this as an Issue over at #364)

**Comment by paulmelis** [comment]
> Out of curiosity, how does the XR-3 system generally behave when controllers are present? Does it disable hand-tracking and switch to controllers automatically or is that a manual process? Or is it neither of those, and just track hands and controllers both at the same time?

I actually don't know for sure. It seems to be application-defined which gets used, as with the included Varjo demos some use hand tracking, while others use the controllers, without any manual config needed. I guess it also depends on how the Unreal/Unity support handles hand tracking, as most applications seem to currently be built with one of those. 

> Also curious, how does XR-3's hand tracking stack up against Quest, if you know? Just wondering if this would be a common request regardless of headset for those demoing to new VR people.

I've only tested the Quest hand tracking on a v1 a bit, plus I haven't checked the latest v2 update. But I suspect there somewhat on the same level, with the Quest even performing better overall. For the XR-3 the major issues (for me) are:

* It's easy for the hand tracking to go out of detection range. E.g. grab an object while looking at it, then rotate your head to see where you want to move the object and tracking might get lost. In that case the hand position somehow resets to the head position, which is really annoying as you then have an object stuck to your head. I'm not sure if this happens in Varjo's OpenXR layer, or maybe is a fallback for SK itself?

* The overlay of virtual fingers with real fingers is not very good, although when I reported this the reply was it was "within limits". There does not seem to be a calibration the user can do to improve this.

* Fat fingers due to incorrect radii, as noted in https://github.com/maluoi/StereoKit/discussions/337. Although this was acknowledged as an issue some weeks ago by the UltraLeap folks, it's unclear when/if this will improve.

**Comment by paulmelis** [comment]
Just noted there's a global override to disable hand tracking in the Varjo Base control software. I can use that for now, but for some applications it's still nice to use the hand tracking

**Comment by paulmelis** [comment]
> The overlay of virtual fingers with real fingers is not very good, although when I reported this the reply was it was "within limits". There does not seem to be a calibration the user can do to improve this.

Just noticed a curious thing in the Varjo dev docs (from https://developer.varjo.com/docs/get-started/hand-tracking#hand-tracking-offset-for-xr-3-and-vr-3):

> When you start to develop with hand tracking, make sure to define an offset for the hand position. This is necessary because the head tracking point for your headset differs from the hand tracking point for Ultraleap.

It then goes on to list XYZ offsets for the XR-3 and VR-2 Pro headsets. I suspect OpenXR will have taken care of these offsets already? As the Z offset listed is almost 10cm, which would definitely be noticeable in my SK app it was off by that much.
